### ISSUE TITLE
ERROR: MethodError: no method matching //(::fmpq_mat, ::fmpq)

### ISSUE THREAD

**Issue by simonbrandhorst** [Nemocas/Nemo.jl#1281]
I would like to have this. It is inconvenient.

**Comment by simonbrandhorst** [comment]
Coming from mathematics I expect that matrix(over a field)/field_element just works .
For instance Magma allows it, sage allows it and I am sure (but did not test) that Mathematics, matlab, maple etc. allow it:
```
> G:= Matrix(Rationals(),1,1,[1]);
> G/2;
[1/2]
```
It is similar to why I prefer 
`2*M`  over `multiply(2, M)` or something similar.

**Comment by simonbrandhorst** [comment]
```
julia> [1 2;]//2
1×2 Matrix{Rational{Int64}}:
 1//2  1//1
```


```

**Comment by wbhart** [comment]
This is not doing what you think it is. It is taking a matrix over the integers and it is creating the matrix over the fraction field with given denominator.

If the // operator is to be used for "division" (since Julia uses / for floating point division) then what operator should we use for constructing the object whose type represents the fraction object?

E.g. should (x^2 + x)//(x + 1) return x or should it return x//1? Those have different types, even if they are mathematical able to be identified.

**Comment by simonbrandhorst** [comment]
For consistency with 
```
julia> 2//2
1//1
```
 It should probably return (x^2 + x)//(x + 1).
But let me be more precise. I propose to set `//` to scalar divison when the base ring is a field. 
Then there is no ambiguity.  But O.K. I see now that this is possibly a change in the current behavior.
```
julia> R,x = PolynomialRing(QQ,"x")
(Univariate Polynomial Ring in x over Rational Field, x)
julia> 2*x
2*x
julia> (2*x)
2*x
julia> (2*x)//2
x
julia> parent((2*x)//2)
Fraction field of Univariate Polynomial Ring in x over Rational Field

julia> (2*x)//QQ(2)
ERROR: MethodError: no method matching //(::fmpq_poly, ::fmpq)
```

**Comment by wbhart** [comment]
It's not a good idea to change the existing behaviour of the system, especially since it would mean introducing a new operator for constructing fractions in that special case.

The existing behaviour has been the behaviour for more than 6 years. I don't believe it is practical to make changes to suit individual taste, especially when that change introduces inconsistency of meaning in the system.

Unfortunately, computers care not just about the mathematical meaning of operations, but also the type. There are very real inconsistencies in other systems because of these sorts of changes, which I've personally long resisted. The different kinds of division must be distinguished in a typed language.

**Comment by wbhart** [comment]
Correct. The coercion principle Tommy outlined is precisely the reason it was designed the way it is. You won't have a consistent computer algebra system if you don't follow that rule. 

It might help to know that this is not arbitrary. I extensively studied the coercion systems of Sage, Magma, Pari/GP and Julia itself about 6 years ago to see what the different possible approaches are before making that very deliberate decision.

The issue becomes particularly acute as the coercion system of the CAS becomes more sophisticated, which it will inevitably.

**Comment by fingolfin** [comment]
The real problem here seems to be that there is no method matching `//(::fmpq_mat, ::fmpq_mat)`. If there was, the example would work, for the same reason `x * y` works when `x` is an `fmpq_mat` and `y` is an `fmpq`.

Yet I can write `x * y^-1` and it works in all combinations of `fmpq` and `fmpq_mat`....

**Comment by thofma** [comment]
For rings (non-fields), `//` creates something in the fraction field. For matrices this of course makes no sense but I don't think it is worthwhile to introduce a "mismatch" between `//(fmpq_poly, ::fmpq)` and `//(fmpq_mat, ::fmpq)`. Sometimes you would stay in the same parent and sometimes not. I do not understand the "yet" part and the "...."? Can you spell out what you mean?

**Comment by wbhart** [comment]
I presume @fingolfin is trying to do this.

```
julia> M = matrix(QQ, [1 2; 3 4])
[1   2]
[3   4]

julia> divexact(M, M)
[1   0]
[0   1]
```
